### PR TITLE
Fix for parsing port at log middleware

### PR DESCRIPTION
**Pull request by spaparaju**
This PR fixes issue#: [4320](https://github.com/thanos-io/thanos/issues/4320)

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Added check for `:` prior to [splitting](https://cs.opensource.google/go/go/+/refs/tags/go1.16.5:src/net/ipsock.go;l=169) the port from the hostURL.

## Verification

Tested on standalone, Kube. environments and did not observe parsing errors in the log files.

### CODE REVIEW

**Review by yeya24 — COMMENTED** [review]
Thanks for the fix.
This pr worth a unit test. Would you mind adding one?

File: pkg/logging/http.go [hunk 1]
```diff
@@ -43,14 +44,24 @@ func (m *HTTPServerMiddleware) HTTPMiddleware(name string, next http.Handler) ht
...
+			_, port, err = net.SplitHostPort(hostPort)
+			if err != nil {
+				level.Error(m.logger).Log("msg", "failed to parse host port for http log decision", "err", err)
+				next.ServeHTTP(w, r)
+				return
+			}
 		}
 
-		decision := m.opts.shouldLog(net.JoinHostPort(r.URL.String(), port), nil)
+		var deciderUrl string = r.URL.String()
```
Comment (squat) [inline pkg/logging/http.go:60]: Nit, we normally use the declare-and-assign operator in go whenever possible
```suggestion
		deciderUrl := r.URL.String()
```

File: pkg/logging/http.go [hunk 2]
```diff
@@ -43,14 +44,24 @@ func (m *HTTPServerMiddleware) HTTPMiddleware(name string, next http.Handler) ht
...
+		if len(port) > 0 {
+			deciderUrl = net.JoinHostPort(r.URL.String(), port)
```
Comment (squat) [inline pkg/logging/http.go:62]: Nit, we can save this call if we reuse the variable
```suggestion
			deciderUrl = net.JoinHostPort(deciderUrl, port)
```

File: pkg/logging/http.go (hunk 1 above)
Comment (squat) [reply pkg/logging/http.go:60]: For consistency with the style guide, we normally capitalize all of the letters in abbreviations:
```suggestion
		var deciderURL string = r.URL.String()
```

**Review by squat — COMMENTED** [review]
Looks good SriKrishna :) i just have small style suggestions

**Comment by bwplotka** [comment]
Nice!

Can we add more clear PR title? It does not tell much what exactly was changed to fix it.

**Comment by spaparaju** [comment]
> Thanks for the fix.
> This pr worth a unit test. Would you mind adding one?

Added a unit test to test with an URL that does not explicitly contain a port number.

**Comment by GiedriusS** [comment]
@spaparaju please fix the DCO:
```
Commit sha: 073971d, Author: Srikrishna Paparaju, Committer: Srikrishna Paparaju; Expected "Srikrishna Paparaju paparaju@gmail.com", but got "Srikrishna Paparaju srikrishna@spaparaj-mac.myfiosgateway.com".

```
https://github.com/thanos-io/thanos/pull/4348/checks?check_run_id=2861409152

**Review by yeya24 — APPROVED** [review]
LGTM

**Comment by ikonyeama** [comment]
Still seeing these errors in the logs after upgrading Thanos from 0.21.1 to 0.24.0 
level=error ts=2022-01-08T07:52:08.191747835Z caller=http.go:48 component=rules protocol=http http.component=server msg="failed to parse host port for http log decision" err="address xyz.com: missing port in address"

**Pull request merged**